skypy/halos/sham.py

```diff
@@ -0,0 +1,1234 @@
+    parent_halo, z_halo = run_file(h_file, 'halo', 'mass', 'z')
+    if print_out:
+        print('Halo catalogues generated in', round((time() - h_st), 2), 's')
+
+    # Generate subhalos and IDs
+    sub_alpha = sub_param[0]  # Vale and Ostriker 2004 mostly
+    sub_beta = sub_param[1]
+    sub_gamma = sub_param[2]
+    sub_x = sub_param[3]
+
+    sub_tim = time()
+    ID_halo, subhalo_m, ID_sub, z_sub = gen_sub_cat(parent_halo, z_halo, sub_alpha,
+                                                    sub_beta, sub_gamma, sub_x)
+
+    if print_out:
+        print('Generated subhalos and IDs in', round((time() - sub_tim), 2), 's')
+        print('')
+
+    # Quench halos and subhalos
+    M_mu = qu_h_param[0]  # Parameters
+    sigma = qu_h_param[1]
+    M_mus = qu_s_param[0]
+    sigmas = qu_s_param[1]
+    baseline_s = qu_s_param[2]
+
+    h_quench = quenching_funct(parent_halo, M_mu, sigma, 0)  # Quenched halos
+    sub_quench = quenching_funct(subhalo_m, M_mus, sigmas, baseline_s)  # Quenched subhalos
+
+    # Galaxy Schechter function parameters
+    rc_param = gal_param[0]
+    rs_param = gal_param[1]
+    bc_param = gal_param[2]
+    bs_param = gal_param[3]
+
+    # Number of halos for each population
+    rc_halo = len(np.where(h_quench == 1)[0])
+    rs_halo = len(np.where(sub_quench == 1)[0])
+    bc_halo = len(np.where(h_quench == 0)[0])
+    bs_halo = len(np.where(sub_quench == 0)[0])
+
+    # Find galaxy mass range (m_star, phi_star, alpha, tag)
+    # TODO Add a way to import a look up table
+    range1 = time()
+    rc_min = find_min(rc_param[0], rc_param[1], rc_param[2], cosmology, z_range, skyarea,
+                      gal_max_h, rc_halo, print_out, run_anyway)
+    if print_out:
+        print('Red central log(minimum mass)', np.log10(rc_min), 'in',
+              round((time() - range1), 4), 's')
+
+    range2 = time()
+    rs_min = find_min(rs_param[0], rs_param[1], rs_param[2], cosmology, z_range, skyarea,
+                      gal_max_s, rs_halo, print_out, run_anyway)
+    if print_out:
+        print('Red satellite log(minimum mass)', np.log10(rs_min), 'in',
+              round((time() - range2), 4), 's')
+
+    range3 = time()
+    bc_min = find_min(bc_param[0], bc_param[1], bc_param[2], cosmology, z_range, skyarea,
+                      gal_max_h, bc_halo, print_out, run_anyway)
+    if print_out:
+        print('Blue central log(minimum mass)', np.log10(bc_min), 'in',
+              round((time() - range3), 4), 's')
+
+    range4 = time()
+    bs_min = find_min(bs_param[0], bs_param[1], bs_param[2], cosmology, z_range, skyarea,
+                      gal_max_s, bs_halo, print_out, run_anyway)
+    if print_out:
+        print('Blue satellite log(minimum mass)', np.log10(bs_min), 'in',
+              round((time() - range4), 4), 's')
+        print('')
+
+    # Get a catalogue for each population
+
+    # Names for created files, randomly generate tag
+    rand_tag = int((10**8)*np.random.rand(1))

```

Comment:
Can this be done more reliably / reproducibly? If it is for differentiating files when multiprocessing, a processor rank could be used instead. Alternatively a time stamp tag could be used too.

Reply:
Possibly removing all of the internal yaml file writing / reading will abrogate the need for this.